core/tx_pool.go

```diff
@@ -1162,3 +1159,68 @@ func (as *accountSet) containsTx(tx *types.Transaction) bool {
 func (as *accountSet) add(addr common.Address) {
 	as.accounts[addr] = struct{}{}
 }
+
+// txLookup is used to track all transactions to allow lookups without contention
+type txLookup struct {
+	all  map[common.Hash]*types.Transaction
+	lock sync.RWMutex
+}
+
+func newTxLookup() *txLookup {
+	return &txLookup{
+		all: make(map[common.Hash]*types.Transaction),
+	}
+}
+
+// calls f on each key and value present in the map
+func (t *txLookup) Range(f func(hash common.Hash, tx *types.Transaction) bool) {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	for key, value := range t.all {
+		if !f(key, value) {
+			break
+		}
+	}
+}
+
+// returns a transaction if it exists in the lookup, or nil if not found
+func (t *txLookup) Get(hash common.Hash) *types.Transaction {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	return t.all[hash]
+}
+
+// returns a transaction if it exists in the lookup, and a bool indicating if it was found
+func (t *txLookup) Find(hash common.Hash) (*types.Transaction, bool) {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	value, ok := t.all[hash]
+	return value, ok
+}
```

Comment:
I don't think we really need this method. `Get` returning `nil` is equivalent to not finding the transaction.